KAFKA-7502: Cleanup KTable materialization logic in a single place

dongjinleekr opened the pull request:
This is a draft cleanup for KAFKA-7502. Here is the details:

1. Make `KTableKTableJoinNode` abstract, and define its child classes (`[NonMaterialized,Materialized]KTableKTableJoinNode`) instead: now, all materialization-related routines are separated into the other classes.
2. `KTableKTableJoinNodeBuilder#build` now instantiates `[NonMaterialized,Materialized]KTableKTableJoinNode` classes instead of `KTableKTableJoinNode`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

dongjinleekr committed:
1. Separate Materialized-related logic into [NonMaterialized|Materialized]KTableKTableJoinNode. 2. Add KTableKTableJoinNodeBuilde#builder.

dongjinleekr committed:
Remove key/value serde decision logic from KTableImpl#doJoin

1. Add KTableKTableJoinNode[Builder]#keySerde.
2. Reorder KTableKTableJoinNodeBuilder's member methods, following the memeber variable definition order.
3. Add KTableKTableJoinNode#[getKeySerde, getValueSerde] methods.
4. Move KTableImpl#[keySerde, valueSerde] decision logic into KTableKTableJoinNodeBuilder#builder and [NonMaterialized, Materialized]KTableKTableJoinNode#[keySerde, valueSerde].

dongjinleekr committed:
Trivial: change method names following coding convention (i.e., remove 'get's.)

dongjinleekr committed:
Remove key/value serde decision logic from KTableImpl#doJoin

1. Make KTableKTableJoinMerger publie. (like KTableReduce, KTableAggregate.)
2. Add KTableKTableJoinMerger#of static methods: works as a workaround to the constructor.
3. Move [internalQueryableName, joinMerge, joinMergeProcessorParameters] variable in KTableImpl#doJoin into [NonMaterialized, Materialized]KTableKTableJoinNode.

dongjinleekr commented:
@guozhangwang Is this what you meant? I found similar routines in `KTableImpl#[filter, map, transform]` methods. If it is correct, I will do the same work to that methods.

Please have a look you are free.

dongjinleekr commented:
retest this please

dongjinleekr commented:
Rebased against the latest trunk. cc/ @guozhangwang @mjsax

guozhangwang reviewed (commented):
@dongjinleekr Thanks for your PR! I went through it and realized my original description on the JIRA ticket may not be very clear. I've hence updated the description and let me know if it clarifies a few things.

As for the PR itself, splitting `KTableKTableJoinNode` seems to me not addressing the original issue, but I do really like the way you inlined `buildJoin`. So how about reducing the scope of this PR to just do the cleanup of inlining `buildJoin`, while we can continue discussing on the ticket at the best solution to achieve the address of the ticket.

dongjinleekr committed:
1. Merge MaterializedKTableKTableJoinNode, NonMaterializedKTableKTableJoinNode into KTableKTableJoinNode. 2. Move all materialization logic in KTableKTableJoinNode into KTable#doJoin.

dongjinleekr committed:
Remove unused fields: KTableKTableJoinNode.KTableKTableJoinNodeBuilder#[builder, valueJoiner]

dongjinleekr commented:
I see. Then, how about this? I rebased against the latest trunk and added two commits (296ddfc, e57ad22); Now, all materialization logic in `KTableKTableJoinNode` is now moved into `KTableImpl#doJoin`. Is this way correct?

dongjinleekr commented:
@guozhangwang I see. Then, how about this? I rebased against the latest trunk and added two commits (296ddfc, e57ad22); Now, all materialization logic in `KTableKTableJoinNode` is now moved into `KTableImpl#doJoin`. Is this way correct?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:
```diff
@@ -495,57 +477,69 @@ public String queryableStoreName() {
+            final Serde<VR> valueSerde = materializedInternal.valueSerde();
+            final String queryableStoreName = materializedInternal.storeName();
+            final StoreBuilder<KeyValueStore<K, VR>> storeBuilder = new KeyValueStoreMaterializer<>(materializedInternal).materialize();
+            kTableJoinNodeBuilder
```
This part and the line 525-529 below can be extracted out of if condition.

dongjinleekr replied:
I tried to, but could not; all parameter variables to `KTableKTableJoinNodeBuilder` must be `final`, but Java does not support conditional instantiation like Scala. 

One alternative approach would be introducing a new class like `MaterializeConfig` to contain them. What do you think?

bbejeck replied:
@dongjinleekr you could do something like:
```java
final Serde<K> keySerde;
final Serde<VR> valueSerde;
final String queryableStoreName;
final StoreBuilder<KeyValueStore<K, VR>> storeBuilder;

if (materializedInternal != null){
   ....
else {
   ....
}

kTableJoinNodeBuilder
       .withKeySerde(keySerde)
       .withValueSerde(valueSerde)
       .withQueryableStoreName(queryableStoreName)
       .withStoreBuilder(storeBuilder);
```
Alternatively you can make them non-`final` and set the default values and update the variables inside the `if` condition, and simply remove the `else` block.  It's a matter of personal style so I'll leave it up to you.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/KTableKTableJoinNode.java:
```diff
@@ -56,9 +64,39 @@
+    }
+
+    public String queryableStoreName() {
+        return queryableStoreName;
```
nit: we do not need to save another reference of queryableStoreName at this class, instead we can add a getter for the field with the same name in `KTableKTableJoinMerger` public and just return `joinMergeProcessorParameters.queryableStoreName()` here.

guozhangwang reviewed (commented):
This PR lgtm. I think this is a good first step towards KAFKA-7502. Left some minor comments.

Call for another review @bbejeck

guozhangwang commented:
@dongjinleekr could you run the unit test locally and see if the error is related?

guozhangwang commented:
retest this please

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/KTableKTableJoinNode.java:
```diff
@@ -85,44 +123,42 @@ public void writeToTopology(final InternalTopologyBuilder topologyBuilder) {
     }
 
-    public static <K, V1, V2, VR> KTableKTableJoinNodeBuilder<K, V1, V2, VR> kTableKTableJoinNodeBuilder() {
+    public static <K, V1, V2, VR> KTableKTableJoinNodeBuilder<K, V1, V2, VR> kTableKTableJoinNodeBuilder(final InternalStreamsBuilder builder) {
```
The `InternalStreamsBuilder` isn't used and can be removed here. The `InternalStreamsBuilder` gets passed in with the `writeToTopology` method.

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/KTableKTableJoinNode.java:
```diff
@@ -131,59 +167,71 @@ private KTableKTableJoinNodeBuilder() {
-                                              otherJoinSide,
-                                              joinThisStoreNames,
-                                              joinOtherStoreNames);
+                null,
```
Why is `valueJoiner` `null` here? Is this intentional?

EDIT: I get it now, the `ValueJoiner` is used in `KTable#doJoin` on lines 483-491 and thus was actually never used by the `KTableKTableJoinNode`.  

Maybe we should alter the constructor of `KTableKTableJoinNode` to not take a `ValueJoiner` at all and explicitly pass  a` null` the parent class on line 60 of `KTableKTableJoinNode`?  

WDYT?

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/KTableKTableJoinNode.java:
```diff
@@ -33,20 +34,27 @@
-    private final MaterializedInternal<K, VR, KeyValueStore<Bytes, byte[]>> materializedInternal;
+    // The name of queryableStore, which stores the join results.
+    private final String queryableStoreName;
+    private final StoreBuilder<KeyValueStore<K, VR>> storeBuilder;
```
Should this get inlined with the types to `StoreBuilder<KeyValueStore<Bytes, byte[]>>`?
\cc @guozhangwang

guozhangwang replied:
Hmm why can we do that? The storeBuilder returned from the `KeyValueStoreMaterializer` is a fully-wrapped one and not an inner bytes-store, right?

bbejeck replied:
Ah, yes that's correct, @dongjinleekr ignore my comment here.

bbejeck reviewed (commented):
Overall this looks good to me, just a couple of minor comments.

bbejeck commented:
Java 8 passed, Java 11 failed but build results already cleaned up.

retest this please

bbejeck commented:
Both Java 8 and Java 11 failed, test results already cleaned out

retest this please

bbejeck commented:
@dongjinleekr what's the status on this PR?  Once the comments are addressed I believe we can merge this PR.

dongjinleekr committed:
1. Remove valueJoiner parameter from KTableKTableJoinNode. 2. Remove KTableKTableJoinNode#queryableStoreName and use KTableKTableJoinMerger#getQueryableName instead.

dongjinleekr committed:
Remove InternalStreamsBuilder parameter from KTableKTableJoinNode#kTableKTableJoinNodeBuilder

dongjinleekr commented:
@bbejeck Here is the update for the `join` operation. I am now working on the other operations and can finish it by this Friday.

guozhangwang commented:
@dongjinleekr Let's use separate PRs on other operations so that we can keep this one small and tangible for reviewing / merging.

BTW the failures are due to compilation errors: `Execution failed for task ':streams:compileJava'.`

Could you run `./gradlew streams:test` locally and see if you can resolve these errors?

dongjinleekr commented:
@guozhangwang Here is the fix. I omitted a file from the last commit. (+1. rebased onto the latest trunk.)

And also, I will separate the other operators from this issue in Jira. Is that okay?

cc/ @bbejeck

dongjinleekr commented:
@guozhangwang Or, only separated PRs for this issue? @bbejeck

bbejeck commented:
Java 8 failed with 
```
org.apache.kafka.common.security.authenticator.SaslAuthenticatorTest.testValidSaslScramSha256
org.apache.kafka.common.security.authenticator.SaslAuthenticatorTest.testMultipleServerMechanisms
kafka.api.UserQuotaTest.testThrottledProducerConsumer
```
Java 11 passed

retest this please

bbejeck reviewed (commented):
This PR looks good to me, less just one minor comment.

bbejeck commented:
>Or, only separated PRs for this issue?

Fine with me to have multiple PRs on one Jira ticket, just make sure to have the Jira number in each PR title and in the PR comments state what parts you are working on.
\cc @guozhangwang

dongjinleekr committed:
Cleanup KTableKTableJoinNode building routine

dongjinleekr commented:
Here it is. OMG, I just learned one more thing about Java's final variable! Thanks @bbejeck :)

And about Issue management, I am +1 to separating the PRs only. In that case, two of the other PRs should be based on the other, since `TableProcessorNode` is used by all of `filter`, `map`, and `transform`. (In other words, one former PR will be the preliminary work for the other two PRs.)

I am now working on `filter` now. Let's roll! :smile:

cc/ @guozhangwang

bbejeck reviewed (approved):
LGTM! Thanks @dongjinleekr

guozhangwang commented:
LGTM!
